src/Campaigns/Blocks/CampaignDonorsBlock/render.php

```diff
@@ -0,0 +1,261 @@
+         */
+        private function fetchCampaign(): ?Campaign
+        {
+            if ( ! isset($this->attributes['campaignId'])) {
+                return null;
+            }
+
+            return give(CampaignRepository::class)->getById($this->attributes['campaignId']);
+        }
+
+        /**
+         * @unreleased
+         */
+        public function render(): void
+        {
+            if ( ! $this->campaign) {
+                return;
+            }
+
+            $query = $this->buildDonorsQuery();
+            $donors = $this->formatDonorsData($query->getAll());
+            $this->renderBlockHtml($donors);
+        }
+
+        /**
+         * @unreleased
+         */
+        private function buildDonorsQuery(): CampaignDonationQuery
+        {
+            $sortBy = $this->attributes['sortBy'] ?? 'top-donors';
+            $query = new CampaignDonationQuery($this->campaign);
+
+            if ($sortBy === 'top-donors') {
+                $query->select(
+                    'donorIdMeta.meta_value as id',
+                    'SUM(amountMeta.meta_value) as amount',
+                    'donors.name as name'
+                )
+                    ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorIdMeta')
+                    ->joinDonationMeta(DonationMetaKeys::AMOUNT, 'amountMeta')
+                    ->leftJoin('give_donors', 'donorIdMeta.meta_value', 'donors.id', 'donors')
+                    ->groupBy('donorIdMeta.meta_value')
+                    ->orderBy('amount', 'DESC');
+            } else {
+                $query->select(
+                    'donation.ID as donationID',
+                    'donorIdMeta.meta_value as id',
+                    'companyMeta.meta_value as company',
+                    'donation.post_date as date',
+                    'amountMeta.meta_value as amount',
+                    'donors.name as name'
+                )
+                    ->joinDonationMeta(DonationMetaKeys::DONOR_ID, 'donorIdMeta')
+                    ->joinDonationMeta(DonationMetaKeys::COMPANY, 'companyMeta')
+                    ->joinDonationMeta(DonationMetaKeys::AMOUNT, 'amountMeta')
+                    ->leftJoin('give_donors', 'donorIdMeta.meta_value', 'donors.id', 'donors')
+                    ->orderBy('donation.ID', 'DESC');
+            }
+
+            $query->limit($this->attributes['donorsPerPage'] ?? 5);
+
+            if ( ! $this->attributes['showAnonymous']) {
+                $query->joinDonationMeta(DonationMetaKeys::ANONYMOUS, 'anonymousMeta')
+                    ->where('anonymousMeta.meta_value', '0');
+            }
+
+            return $query;
+        }
+
+        /**
+         * @unreleased
+         */
+        private function formatDonorsData(array $donors): array
+        {
+            return array_map(static function ($entry) {
+                if (isset($entry->date)) {
+                    $entry->date = Carbon::parse($entry->date)->diffForHumans();
```

Comment:
@pauloiankoski we have to be careful about introducing php libraries because they can conflict with other plugins.  this is the reason we Struass.  However, do we even need this one? we have functions in WP like [human_time_diff](https://developer.wordpress.org/reference/functions/human_time_diff/) and our own [Temporal](https://github.com/impress-org/givewp/blob/3fe9f4dd7f82cefe6c41c37df9e00713dba10c28/src/Framework/Support/Facades/DateTime/TemporalFacade.php#L15) class

Reply:
Thanks for sharing that function. I do agree with you, btw.

Refactored on 60aa32e0f8d54b9d0b3e375bd22afe52de934953